druid/src/widget/scrollbar.rs

```diff
@@ -0,0 +1,375 @@
+use std::mem;
+
+use crate::kurbo::{Rect, RoundedRect, Size};
+use crate::theme;
+use crate::{
+    BoxConstraints, Data, Env, Event, EventCtx, LayoutCtx, PaintCtx, RenderContext, Selector,
+    UpdateCtx, Widget,
+};
+
+/// Trait used for shared state between
+/// the scrollbar and the scroll container.
+pub trait ScrollControlState: Data + PartialEq {
+    /// The last position reported by a MouseMove event.
+    /// used to calculate the deltas without maintaining
+    /// an offset.
+    fn last_mouse_pos(&self) -> f64;
+    /// The identifier for this state object
+    fn id(&self) -> u64;
+    /// The size of a page. Used to calculate
+    /// scroll distances when clicking on the
+    /// scroll track
+    fn page_size(&self) -> f64;
+    /// The maximum distance in pixels to scroll
+    fn max_scroll_position(&self) -> f64;
+    /// The minimum distance in pixels to scroll
+    fn min_scroll_position(&self) -> f64;
+    /// boolean for enabling/disabling the mouse
+    fn mouse_wheel_enabled(&self) -> bool;
+    /// scrollbar travel distance / scroll container travel distance
+    fn scale(&self) -> f64;
+    /// The current scroll position. This value
+    /// is always between min and max scroll positions
+    fn scroll_position(&self) -> f64;
+    /// true when tracking mouse move events
+    fn tracking_mouse(&self) -> bool;
+
+    fn set_last_mouse_pos(&mut self, val: f64);
+    fn set_page_size(&mut self, val: f64);
+    fn set_max_scroll_position(&mut self, val: f64);
+    fn set_min_scroll_position(&mut self, val: f64);
+    fn set_mouse_wheel_enabled(&mut self, val: bool);
+    fn set_tracking_mouse(&mut self, val: bool);
+    fn set_scale(&mut self, val: f64);
+    /// Sets the raw value for the scroll position
+    /// without respecting min and max. Use this
+    /// for pull or bounce calculations on the container
+    fn set_scroll_position(&mut self, val: f64);
+
+    fn set_scroll_pos_from_delta(&mut self, delta: f64) {
+        let scroll_position = self.scroll_position() + delta;
+
+        self.set_scroll_position(
+            self.min_scroll_position()
+                .max(scroll_position.min(self.max_scroll_position())),
+        );
+    }
+}
+
+struct AnimationState {
```

Comment:
A time-based animation struct. This should be replaced with something more robust later but seems to perform well for the time.